Publishing from Web Applications

DavidBoike opened the pull request:
Here's new doco for publishing from web applications. My plan would be to have this as authoritative, and not rewrite my own blog piece, but put a big header at the top of it saying it's wildly outdated information with a big link to this article.

Looking for input from @Particular/engineers and @Particular/documentation 

My experience is entirely MSMQ, so I'm sure I'm saying things about the other transports that are conceptually right (or maybe it's not) but the wording is just wrong. Please comment or even just add commits.

There are placeholders for 2 diagrams. If I can get buy-in on the text and the idea behind them, I'll add the diagrams.

Also, the PR doesn't include this article being linked in from anywhere yet.

SzymonPobiega commented:
Great! That clears a lot of things to me.

DavidBoike commented:
Thanks @SzymonPobiega! Any problems with the phrasing and how it applies to the non-MSMQ transports?

Would love similar feedback from @yvesgoeleven, @andreasohlund, @johnsimons since you hold the GitHub additions records for Azure (Yves), Rabbit (Andreas), SqlServer (John)

andreasohlund commented on nservicebus/architecture/publishing-from-web-applications.md:
```diff
@@ -0,0 +1,89 @@
+redirects:
+---
+
+You may have heard that [it's bad practice to publish events from a web application](http://www.make-awesome.com/2010/10/why-not-publish-nservicebus-messages-from-a-web-application/). This guidance is a bit outdated, as improvements Microsoft has made to IIS, more varied methods of hosting .NET web applications, and the addition of additional NServiceBus transports have made things much more nuan ...
```
Do we need to link to your blog here? It feels weird to me that official docs use phrasing like "you may have heard". Should we just write what the guidance is as of now? (of course taking different transports into account as needed)

DavidBoike replied:
I wrestled with that. Having a reciprocal link doesn't hurt SEO for either page, given that page on my blog comes up first on a bunch of different searches. In my mind. having the link prevents people from going "Wait, but I thought..." because it illustrates that before, X, and now, X agrees that the guidance is Y. But we could take that chunk out for sure. What do you think?

Either way my blog post will be changed to link to the official doco right at the top.

andreasohlund commented on nservicebus/architecture/publishing-from-web-applications.md:
```diff
@@ -0,0 +1,89 @@
+
+Historically, there have been good reasons to avoid publishing events from within web applications:
+
+1. **Transactions and Consistency** - Generally good events should announce that something has *already* happened. Many times this involves making some changes to a database and then publishing the result. But HTTP is inherently unreliable and does not have built-in retries. If an exception occurs before the event is published, you may lose your only opportunity to publish that event. In these ci ...
```
Should be firmer and say that event `should` announce something that has happened?

Should we link to http://docs.particular.net/nservicebus/messaging/messages-events-commands ?

andreasohlund replied:
Could "with the payload of the HTTP request" be misinterpreted as "dump the http headers and body into a message" ? (or am I overthinking it)

SeanFeldman replied:
I understood "payload" as information you'd normally post over http.

andreasohlund commented on nservicebus/architecture/publishing-from-web-applications.md:
```diff
@@ -0,0 +1,89 @@
+Historically, there have been good reasons to avoid publishing events from within web applications:
+
+1. **Transactions and Consistency** - Generally good events should announce that something has *already* happened. Many times this involves making some changes to a database and then publishing the result. But HTTP is inherently unreliable and does not have built-in retries. If an exception occurs before the event is published, you may lose your only opportunity to publish that event. In these ci ...
+2. **Web Application Scale-Out** - Because events can only be published from one logical publisher, it can be problematic to scale out a web application that is publishing events. More on this later.
```
Perhaps out of scope for this one but we need to explain why there can be only one logical publisher and link to that page. (http://docs.particular.net/nservicebus/messaging/publish-subscribe/ doesn't seem to cover it)

DavidBoike replied:
You saying we need a new page for that? Or update that page you linked? BTW that page you linked seems to be more about the distributor than anything else.

andreasohlund replied:
Agree, we need a new page

andreasohlund commented on nservicebus/architecture/publishing-from-web-applications.md:
```diff
@@ -0,0 +1,89 @@
+
+1. **Transactions and Consistency** - Generally good events should announce that something has *already* happened. Many times this involves making some changes to a database and then publishing the result. But HTTP is inherently unreliable and does not have built-in retries. If an exception occurs before the event is published, you may lose your only opportunity to publish that event. In these ci ...
+2. **Web Application Scale-Out** - Because events can only be published from one logical publisher, it can be problematic to scale out a web application that is publishing events. More on this later.
+3. **Web App Recycling** - Prior to IIS 7.5 introducing the AlwaysRunning feature (AKA AlwaysOn), web applications not receiving could be recycled or idled at any time, which would result in incoming messages building up in an input queue, without an application actively processing them.
```
"web applications not receiving" doesn't read well for me. Change to "Idle web applications could be shutdown or recycled at any time" ?

Do we need to split shutdown vs recycled since they are associated with different problems:

Shutdown: Messages pile up since there is no receiving process running
Recycle: You might "miss" message functionally to be triggered of you have callbacks or publishing over signalr. Is recycle even relevant for the "publishes" ? (I'd say only if the publish is over signalr)

Actually is the relevant at all? Or should this article be a more generic "Dangers when receiving from a webapp"?

Or is that a new different guidance page?

SeanFeldman commented on nservicebus/architecture/publishing-from-web-applications.md:
```diff
@@ -0,0 +1,89 @@
+
+Some transports (Azure Service Bus, RabbitMQ) support publish/subscribe natively. This means that when an endpoint wants to subcribe to an event, it contacts the broker directly, and the broker maintains the list of subscriptions.
+
+Other transports (MSMQ, SQL, Azure Storage Queues) do not support native publish/subscribe, and instead use storage-driven publishing with message-driven subscriptions. This means that each endpoint is responsible for maintaining its own subscription storage, usually in a database. When an endpoint wants to subscribe to an event, it sends a subscription request message to the owner endpoint, whic ...
```
Could it read nicer if order changed to be "...and instead use message-driven subscriptions with storage-driven publishing."
It's probably my head insisting on pub/sub order :)

andreasohlund replied:
How about "For transports that lack native pub/sub capabilities NServiceBus will provide you with similar semantics by sending subscription requests to relevant publishers which then maintains this list in a subscription storage..." ?

SeanFeldman replied:
Like that.

SeanFeldman commented on nservicebus/architecture/publishing-from-web-applications.md:
```diff
@@ -0,0 +1,89 @@
+
+The publishing process also differs based on the message transport.
+
+A transport that supports native publish/subscribe (Azure Service Bus, RabbitMQ) will send the event message directly to the broker, addressed to a topic, which will distribute the message to all interested parties.
```
Should "interested parties" be subscribers?

andreasohlund replied:
Just remove "addressed to a topic"?  since that's broker specific. In Rabbit "topics" are called Exchanges

andreasohlund commented on nservicebus/architecture/publishing-from-web-applications.md:
```diff
@@ -0,0 +1,89 @@
+
+## Publish/Subscribe Mechanics
+
+First it's important to understand how publish/subscribe actually works, consisting of the subscription phase, and then the actual publish operation.
```
We have some doco on this that we should be able to link to eg. http://docs.particular.net/nservicebus/messaging/publish-subscribe/

DavidBoike replied:
That article talks about how publishing works with the distributor more than anything else. I don't think that's a good link for this case. (And maybe an article that should be revisited/recategorized/retitled as well?)

andreasohlund commented on nservicebus/architecture/publishing-from-web-applications.md:
```diff
@@ -0,0 +1,89 @@
+
+An endpoint must register its interest in a message by subscribing to a specific message type. How this occurs is dependent upon the message transport being used.
+
+Some transports (Azure Service Bus, RabbitMQ) support publish/subscribe natively. This means that when an endpoint wants to subcribe to an event, it contacts the broker directly, and the broker maintains the list of subscriptions.
```
"list of subscriptions" sound a bit impl specific. How about "keeps track of subscribers for each event" ?

andreasohlund commented on nservicebus/architecture/publishing-from-web-applications.md:
```diff
@@ -0,0 +1,89 @@
+
+A transport that uses storage-driven publishing will act differently. When it needs to publish, it will first consult its subscription storage to determine the interested parties. It will then send an independent copy of the event message to each subscriber.
+
+While it is common to think of an event being published *from* a specific location, this doesn't always reflect the reality of the messaging mechanics. For native pub/sub brokers, the "publish from" location (the `Endpoint` value in the `MessageEndpointMappings` section) relates to a Topic in the broker.
```
Not correct, you need no mappings at all for Pub/Sub to work with the brokers since you just need to know the event type. The broker then figures out the rest

andreasohlund commented on nservicebus/architecture/publishing-from-web-applications.md:
```diff
@@ -0,0 +1,89 @@
+
+## Load Balancing
+
+Typically, web applications will be scaled out with a network load balancer either to handle large amounts of traffic, or for high availability, or both. This is further complicated by elastic cloud scenarios, where additional instances of a web role can be provisioned and later removed, either manually or automatically, to keep up with a changing amount of load.
```
Move this to the new doco page like we discussed since it's not relevant for pub/sub in web apps?

DavidBoike replied:
I think you're on the wrong paragraph? This is still important because it sets up the conditions a scaled out web app has to satisfy to all publish as one logical publisher.

Right?

andreasohlund commented on nservicebus/architecture/publishing-from-web-applications.md:
```diff
@@ -0,0 +1,89 @@
+It can be useful, at times, to publish events from a web application. Always keep the following guidelines in mind.
+
+* Be mindful of transport differences
+    * For transports using native pub/sub must publish (Azure Service Bus, RabbitMQ), all web application instances must publish to the same topic.
```
Not relevant, just remove the bullet point?

DavidBoike commented:
I currently have this article in /nservicebus/architecture/publishing-from-web-applications.md.

Should it be in /nservicebus/messaging instead?

/cc @SimonCropp

andreasohlund commented on nservicebus/architecture/publishing-from-web-applications.md:
```diff
@@ -0,0 +1,90 @@
+In the case of storage-driven transports, however, the "publish from" location actually relates to the input queue that is set up to receive subscription requests.
+
+## Load Balancing
+
```
The load balancing aspect doesn't seem that relevant to me? Yes, you need to share the same subscription storage but the same applies to eg. workers behind a load balancer?

DavidBoike replied:
OK, how about kill that section, and then add an additional paragraph to the "Publishing" section above it, tacking on to the end of the talking about storage-driven transports, to say something like "This means that for storage-driven transports to successfully publish as one logical.......must all share the same subscription storage. This is true of all NServiceBus endpoints, whether they are web applications deployed behind a network load balancer, or scaled-out worker endpoints."

DavidBoike replied:
That really didn't work. The section needs to exist so that it leads into the Storage-driven topology section next. Did my best to indicate that hey, this isn't really any different than service endpoints.

SimonCropp commented:
@DavidBoike yeah this shouldnt be in `architecture`

DavidBoike commented:
Moved the article to the /nservicebus/hosting directory, tweaked the load balancing section, created and annotated the diagram. Is this done?

SimonCropp commented:
@DavidBoike do you want it to be discoverable from the menu https://github.com/Particular/docs.particular.net#menu ?

andreasohlund commented:
Looks good to go

SimonCropp commented on nservicebus/hosting/publishing-from-web-applications.md:
```diff
@@ -0,0 +1,91 @@
+
+For storage-driven transports, it is inadvisable to have one of the web applications receive subscription requests. Instead, each web application instance can be implemented as an `ISendOnlyBus`, and a back-end service endpoint can be responsible for receiving the subscription request messages and updating the subscription storage.
+
+![Storage-driven transport publishing topology](storage-based-publish-topology.png)
```
btw we support image resizing https://github.com/Particular/docs.particular.net#image-sizing

DavidBoike commented:
Added the width and added to menu under Hosting, although this makes me re-skeptical about its proper placement in the Hosting category. But better to get it out than bikeshed all day.

@SimonCropp do you want to do the honors and merge?
